### ISSUE TITLE
Support for Unity Mathematics

### ISSUE THREAD

**Issue by provencher** [FreyaHolmer/Mathfs#13]
Just came across this repo, and it looks fantastic!
I've been in the process of more heavily investing in Burst optimizing static functions, and this package looks largely static to begin with.

Any chance it could leverage mathematics utilities and structs instead of the ones in the old math library, so that they could automatically be optimized using the compiler optimizations supported by the new math library?

**Comment by FreyaHolmer** [comment]
I'm really torn on this one, since it seems like, well, unity itself is torn between their old math structs vs the new ones. I don't know what I should do!

**Comment by provencher** [comment]
Well with IL2CPP the new math library is pretty much equivalent to the old one, but some of the older components still require Vector3s and Quaternions, instead of float3 and quaternions, but you can easily cast between them.

Also, if you work with native arrays arrays of floatxs from the new library (starting with unity 2022.1), you can reinterpret cast between them and the vector equivalents at no cost. Burst works on static functions too now, so you can even make the whole thing burst optimized.

See here on burst support!
https://docs.unity3d.com/Packages/com.unity.burst@1.6/manual/docs/CSharpLanguageSupport_Types.html

In any case, this is just a feature request, I've found gains of up to 50% using burst optimized math over the traditional approach.

**Comment by Logyrac** [comment]
I know this is old but I have to concur here. Especially with Unity moving towards their ECS for all path, where ECS will be baked into the core engine and GameObjects will have underlying ECS logic replacing much of the current Behaviour components for built-in types (such as unification of ECS Transform and Behaviour Transform), the ECS ecosystem uses Unity.Mathematics almost exclusively as it's generally better for Burst compilation and optimized for SIMD. Effectively this means a large amount of core Unity functionality will start to rely more heavily on Unity.Mathematics down the road, which likely means the newer Mathematics package will get more attention in the long run than the older types.

Considering that Vector2/3 and float2/3 have implicit casts and can also be used with reinterpret_cast switching to Unity.Mathematics would still largely allow the use of Vector2/Vector3/Vector4/Quaternion/etc without much hassle. In the long run it is probably better to migrate over to Unity.Mathematics, though given the size of this project that may be more of an undertaking than it seems at first. Types can't simply be swapped out everywhere, certain utility functions like ToAngleAxis on quaternions don't exist and have to be implemented, and float3 doesn't have the equivalents of Vector3.forward and etc. Though it may be possible to remove the need for types like Vector2Matrix4x1 and use the float4x2 matrix types (and equivalents), and there may be opportunities for optimizing for SIMD or vectorizing certain operations.

**Comment by FreyaHolmer** [comment]
yeah! I've actually started doing this personally for most of my projects, I just haven't committed/cleaned it up enough to push it here, but I probably should soon :)
